src/python/pants/vcs/changed.py

```diff
@@ -43,15 +44,26 @@ async def find_changed_owners(request: ChangedRequest) -> ChangedAddresses:
     owners = await Get(Owners, OwnersRequest(request.sources, filter_by_global_options=True))
     if request.dependees == DependeesOption.NONE:
         return ChangedAddresses(owners)
-    dependees_with_roots = await Get(
+
+    # See https://github.com/pantsbuild/pants/issues/15313. We filter out target generators because
+    # they are not useful as aliases for their generated targets in the context of
+    # `--changed-since`. Including them makes it look like all sibling targets from the same
+    # target generator have also changed.
+    #
+    # However, we also must be careful to preserve if target generators are direct owners, which
+    # happens when a generated file is deleted.
+    owner_target_generators = FrozenOrderedSet(
+        addr.maybe_convert_to_target_generator() for addr in owners if addr.is_generated_target
+    )
+    dependees = await Get(
         Dependees,
         DependeesRequest(
             owners,
             transitive=request.dependees == DependeesOption.TRANSITIVE,
-            include_roots=True,
+            include_roots=False,
         ),
     )
-    return ChangedAddresses(dependees_with_roots)
+    return ChangedAddresses(FrozenOrderedSet(owners) | (dependees - owner_target_generators))
```

Comment:
The `include_roots=False` and then re-including them below is to handle the case where a `BUILD` file changes, and so a generator is directly marked changed...? Might be good to comment on that.

Reply:
That's what this is meant to say:

```
# However, we also must be careful to preserve if target generators are direct owners, which
# happens when a generated file is deleted.
```